src/Ocelot/Multiplexer/MultiplexingMiddleware.cs

```diff
@@ -1,219 +1,265 @@
+        JToken jObject, HttpContext httpContext, DownstreamRoute downstreamRoute)
+    {
+        var tasks = new List<Task<HttpContext>>();
+        var values = jObject.SelectTokens(matchAdvancedAgg.JsonPath).Select(s => s.ToString()).Distinct();
+        foreach (var value in values)
+        {
+            var tPnv = httpContext.Items.TemplatePlaceholderNameAndValues();
+            tPnv.Add(new PlaceholderNameAndValue('{' + matchAdvancedAgg.Parameter + '}', value));
+            tasks.Add(ProcessRouteAsync(httpContext, downstreamRoute, _next, tPnv));
         }
 
-        private static HttpContext Copy(HttpContext source)
-        {
-            var target = new DefaultHttpContext();
+        return tasks;
+    }
 
-            foreach (var header in source.Request.Headers)
-            {
-                target.Request.Headers.TryAdd(header.Key, header.Value);
-            }
+    /// <summary>
+    /// Process a downstream route asynchronously.
+    /// </summary>
+    /// <returns>The cloned Http context.</returns>
+    private static Task<HttpContext> ProcessRouteAsync(HttpContext sourceContext, DownstreamRoute route,
+        RequestDelegate next, List<PlaceholderNameAndValue> placeholders = null)
+    {
+        var newHttpContext = CreateThreadContext(sourceContext);
+        CopyItemsToNewContext(newHttpContext, sourceContext, placeholders);
+        newHttpContext.Items.UpsertDownstreamRoute(route);
 
-            target.Request.Body = source.Request.Body;
-            target.Request.ContentLength = source.Request.ContentLength;
-            target.Request.ContentType = source.Request.ContentType;
-            target.Request.Host = source.Request.Host;
-            target.Request.Method = source.Request.Method;
-            target.Request.Path = source.Request.Path;
-            target.Request.PathBase = source.Request.PathBase;
-            target.Request.Protocol = source.Request.Protocol;
-            target.Request.Query = source.Request.Query;
-            target.Request.QueryString = source.Request.QueryString;
-            target.Request.Scheme = source.Request.Scheme;
-            target.Request.IsHttps = source.Request.IsHttps;
-            target.Request.RouteValues = source.Request.RouteValues;
-            target.Connection.RemoteIpAddress = source.Connection.RemoteIpAddress;
-            target.RequestServices = source.RequestServices;
-            target.RequestAborted = source.RequestAborted;
-            target.User = source.User;
-            return target;
-        }
+        return Fire(newHttpContext, next);
+    }
+
+    /// <summary>
+    /// Copying some needed parameters to the Http context items.
+    /// </summary>
+    private static void CopyItemsToNewContext(HttpContext target, HttpContext source,
+        List<PlaceholderNameAndValue> placeholders = null)
+    {
+        target.Items.Add("RequestId", source.Items["RequestId"]);
```

Comment:
Is `RequestId` defined in docs?
Anyway, define a string constant in the code!

Reply:
Ok, it's used in the logs too.